compose: fix environment interpolation from the client

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Fix (newly support?) environment interpolation from the client

e.g.
```yaml
services:
  x:
    environment:
# needs to be propagated from the client
      - FOO

```

**- How I did it**

Please refer to the code.

**- How to verify it**
Prepare following files:

`docker-compose.yaml`:
```yaml
version: "3"
services:
  x:
    image: "busybox"
    command: ["sh", "-c", "env; tail -f /dev/null"]
    environment:
      - FOO
      - BAR
      - QUX
      - QUUX=quux_from_yaml
    env_file: a.env
```

`a.env`:

```
FOO=foo_from_file
BAR=bar_from_file
BAZ=baz_from_file
```

on Docker Compose:

```console
$ FOO=foo_from_env QUUX=quux_from_env docker-compose up
```

on Docker Swarm-mode:

```
$ FOO=foo_from_env QUUX=quux_from_env docker stack deploy -c docker-compose.yaml stack01
$ docker service logs stack01_x
```


Result:

Composer | `FOO` | `BAR` | `BAZ`|`QUX`|`QUUX`
--- | --- | --- | ---|---|---
Compose 1.11.1, <br> Swarm-mode with this PR | `"foo_from_env"` | unset (not `""`) | `"baz_from_file"` | unset (not `""`) | `"quux_from_yaml"`
Swarm-mode 1.13.1 | `""` | `""` | `"baz_from_file"`|`""`|`"quux_from_yaml"`


~~For `BAR`, I wonder Compose's result (unset) is unexpected and should be `""`, but not sure. (Could not find any doc :sweat_smile:)~~


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
compose: support environment interpolation from the client

**- A picture of a cute animal (not mandatory but encouraged)**

![446374156](https://cloud.githubusercontent.com/assets/9248427/22686242/aa37456a-ed66-11e6-9d9c-594516861df1.png)

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda committed:
compose: fix environment interpolation from the client

For an environment variable defined in the yaml without value,
the value needs to be propagated from the client, as in Docker Compose.

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda commented:
cc @dnephin

dnephin commented on cli/compose/loader/loader.go:
```diff
@@ -380,9 +393,13 @@ func resolveVolumePaths(volumes []string, workingDir string) error {
+		home, ok := lookupEnv("HOME")
+		if !ok {
+			panic(errors.New("cannot expand '~', because the environment lacks HOME"))
+		}
```
I don't think this should be a panic. At most a warning.

AkihiroSuda replied:
@dnephin What should we do with warning?
Keep unparsed '~' as-is? Or Ignore the entry?

dnephin replied:
Keep it as-is with the `~` I think

dnephin commented on cli/compose/loader/loader.go:
```diff
@@ -358,10 +363,18 @@ func resolveEnvironment(serviceConfig *types.ServiceConfig, workingDir string) e
 
 	serviceConfig.Environment = environment
```
nit: To be consistent I think this line should be at the end of the function, right before `return nil`

dnephin commented on cli/compose/loader/loader.go:
```diff
@@ -358,10 +363,18 @@ func resolveEnvironment(serviceConfig *types.ServiceConfig, workingDir string) e
 
+	// env is prioritized over the file
+	for k := range serviceConfig.Environment {
+		v, ok := lookupEnv(k)
```
I think this is missing a check to see if `v == ""`. If `v` is not the empty string we should not override from the client environment.

dnephin reviewed (commented):
This looks great, thanks!
 
Just a couple small issues

aaronlehmann commented on cli/command/stack/deploy.go:
```diff
@@ -180,6 +180,13 @@ func getConfigDetails(opts deployOptions) (composetypes.ConfigDetails, error) {
+	env := os.Environ()
+	details.Environment = make(map[string]string, len(env))
+	for _, s := range env {
+		// if value is empty, s is like "K=", not "K".
```
This is probably a reasonable assumption, but it couldn't hurt to check.

AkihiroSuda commented:
@dnephin @aaronlehmann 
Updated PR (sorry for being late)

Now this PR is much more compatible with Docker Compose, but maybe breaking compatibility with Docker Swarm-mode 1.13.1. (Please refer to changes in UT)

cc @shin-

dnephin commented on cli/compose/loader/loader.go:
```diff
@@ -352,20 +357,34 @@ func resolveEnvironment(serviceConfig *types.ServiceConfig, workingDir string) e
+			if ok {
+				environment[k] = interpolatedV
+			} else {
+				delete(environment, k)
```
I don't think you ever want to delete things from the environment. What's this about?

AkihiroSuda replied:
If the environment variable is specified to be interpolated in the YAML but its value is unset in the actual environment, it should be deleted (in Docker Compose's implementation).

Please refer to the example table listed in the description text of this PR.

dnephin replied:
I don't think it's actually deleted. I don't see that in the Compose code anywhere. I think that's handled by the server when starting a container. It creates the variable unset.

AkihiroSuda replied:
Yes, you are right, I didn't know that 😅 

Opened https://github.com/docker/docker/pull/31634 for clarifying this behavior in the doc.

I'll update this PR as well in short.

dnephin commented on cli/compose/loader/loader.go:
```diff
@@ -352,20 +357,34 @@ func resolveEnvironment(serviceConfig *types.ServiceConfig, workingDir string) e
+				environment[k] = interpolatedV
+			} else {
+				environment[k] = v
+			}
```
Here's the logic from Compose: https://github.com/docker/compose/blob/master/compose/config/config.py#L589-L598

I think we can reduce the size of this function quite a bit by matching the logic.

Once `delete()` is removed from below, i think these blocks will be nearly identical. So we could have a function like this:

```go

func updateEnvironment(environment, map[string]string, vars map[string]string, lookupEnv template.Mapping) map[string[string {
    ...
}

environment = updateEnvironment(environment, runconfigopts.ConvertKVStringsToMap(envVars), lookupEnv)
serviceConfig.Environment = updateEnvironment(environment, serviceConfig.Environment, lookupEnv)
return nil

```

AkihiroSuda replied:
Thank you for the pointer, I'll look into that.

vdemeester reviewed (commented):
Design looks good :)

cpuguy83 commented:
Ping, any update here?

AkihiroSuda commented:
updated PR, sorry for delay

AkihiroSuda commented:
oops, still something is wrong... (expected: `BAR`, got `BAR=`)

```
$ docker inspect task_container_of_the_example_above
            "Env": [
                "BAR=",
                "BAZ=baz_from_file",
                "FOO=foo_from_env",
                "QUUX=quux_from_env",
                "QUX=",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
```

AkihiroSuda commented:
trying to fix the issue 👆 by defining `MappingWithEquals` as `map[string]*string` rather than `map[string]string`, but getting stuck due to reflection ... 😭  https://github.com/AkihiroSuda/docker/commits/fix-stack-env.wip.20170308

dnephin committed:
Fix environment resolving.

Load from env should only happen if the value is unset.
Extract a buildEnvironment function and revert some changes to tests.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
I've rebased this PR, and fixed the remaining issue with empty values.

AkihiroSuda commented:
@dnephin Thank you a lot :smiley:

AkihiroSuda commented on cli/compose/types/types.go:
```diff
@@ -135,7 +135,10 @@ type StringOrNumberList []string
 
 // MappingWithEquals is a mapping type that can be converted from a list of
 // key=value strings
```
```go
// MappingWithEquals is a mapping type that can be converted from a list of
// key[=value] strings.
// For the key with an empty value (`key=`), the mapped value is set to a pointer to `""`.
// For the key without value (`key`), the mapped value is set to nil.
```

AkihiroSuda replied:
committed as 2cdc9c42eb138bcaffb1cbd4172bd18b69446394

vdemeester reviewed (approved):
LGTM 🦁

AkihiroSuda committed:
compose: update the comment about MappingWithEquals

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda commented:
LGTM for @dnephin 's work 😃

dnephin reviewed (approved):
LGTM

AkihiroSuda commented:
@vieux @dnephin 

I'm unsure this is a new feature of bug-fix :sweat_smile:, but can we cherry-pick this for 17.04?

thaJeztah commented:
ping @AkihiroSuda could you prepare a cherry-pick for this?

AkihiroSuda commented:
sure, opened https://github.com/docker/docker/pull/32033

bitsofinfo commented:
When is this slated to be released?

thaJeztah commented:
This has been released quite a while ago already

thaJeztah commented:
Docker 17.04 and up should have this change
